src/wamr/filesystem.cpp

```diff
@@ -1,20 +1,432 @@
+
+    return doFileStat(fd, "", statApp);
+}
+
+static int32_t wasi_fd_prestat_dir_name(wasm_exec_env_t exec_env,
+                                        int32_t fd,
+                                        char* path,
+                                        int32_t* pathLen)
+{
+    SPDLOG_DEBUG("S - fd_prestat_dir_name {}", fd);
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    if (!module->getFileSystem().fileDescriptorExists(fd)) {
+        return __WASI_EBADF;
+    }
+
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+    const std::string& pathStr = fileDesc.getPath();
+
+    VALIDATE_NATIVE_ADDR(path, pathStr.size());
+    std::copy(pathStr.begin(), pathStr.end(), path);
+
+    return __WASI_ESUCCESS;
+}
+
+typedef struct wasi_prestat_app
+{
+    __wasi_preopentype_t pr_type;
+    uint32_t pr_name_len;
+} wasi_prestat_app_t;
+
+static int32_t wasi_fd_prestat_get(wasm_exec_env_t exec_env,
+                                   int32_t fd,
+                                   wasi_prestat_app_t* prestatApp)
+{
+    SPDLOG_DEBUG("S - fd_prestat_get {}", fd);
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    if (!module->getFileSystem().fileDescriptorExists(fd)) {
+        return __WASI_EBADF;
+    }
+
+    VALIDATE_NATIVE_ADDR(prestatApp, sizeof(wasi_prestat_app_t));
+
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+
+    prestatApp->pr_type = fileDesc.wasiPreopenType;
+    prestatApp->pr_name_len = fileDesc.getPath().size();
+
+    return __WASI_ESUCCESS;
+}
+
+typedef struct iovec_app
+{
+    uint32_t buffOffset;
+    uint32_t buffLen;
+} iovec_app_t;
+
+static int32_t wasi_fd_read(wasm_exec_env_t exec_env,
+                            int32_t fd,
+                            iovec_app_t* ioVecBuffApp,
+                            int32_t ioVecCountApp,
+                            int32_t* bytesRead)
+{
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    storage::FileSystem& fileSystem = module->getFileSystem();
+    std::string path = fileSystem.getPathForFd(fd);
+
+    SPDLOG_DEBUG("S - fd_read {} ({})", fd, path);
+
+    storage::FileDescriptor fileDesc = fileSystem.getFileDescriptor(fd);
+
+    // Translate app iovecs to native ones
+    auto ioVecBuffNative = new ::iovec[ioVecCountApp];
+    for (int i = 0; i < ioVecCountApp; i++) {
+        VALIDATE_APP_ADDR(ioVecBuffApp[i].buffOffset,
+                          sizeof(char) * ioVecBuffApp[i].buffLen);
+
+        ::iovec ioVecNative{
+            .iov_base = ADDR_APP_TO_NATIVE(ioVecBuffApp[i].buffOffset),
+            .iov_len = ioVecBuffApp[i].buffLen,
+        };
+
+        ioVecBuffNative[i] = ioVecNative;
+    }
+
+    // Read from fd
+    VALIDATE_NATIVE_ADDR(bytesRead, sizeof(int32_t))
+    *bytesRead = ::readv(fileDesc.getLinuxFd(), ioVecBuffNative, ioVecCountApp);
+
+    delete[] ioVecBuffNative;
+
+    return __WASI_ESUCCESS;
```

Comment:
Same comment about duplicating the logic for this function from the existing WAVM implementation, could we refactor it into the `FileSystem` class?